### ISSUE TITLE
Scrollable windows?

### ISSUE THREAD

**Issue by cblc** [libui-ng/libui-ng#200]
When creating forms that can be longer than the display height, it would be great to be able to create them inside automatically scrollable windows. Can this be done currently? (I guess no, because I saw this PR: https://github.com/andlabs/libui/pull/534). Is this planned?

I believe this addition would be very powerful, because it would let to use libui in a similar fashion to AntTweakBar for example (ie: just define the form fields you need, without caring about whether they fit or not in a window).

**Comment by cody271** [comment]
> Can this be done currently? (I guess no, because I saw this PR: [andlabs/libui#534](https://github.com/andlabs/libui/pull/534)). Is this planned?

Yes a scrollable container control is definitely needed.

That PR looks like a good start, with working Unix backend and an example program.
Someone will have to write the Mac and Windows backends to finish it. PRs welcome!

**Comment by cody271** [comment]
Original issue: https://github.com/andlabs/libui/issues/178

**Comment by cblc** [comment]
I see that darwin/form.m uses NSScrollView. Does this mean the MacOS backend already has a bit of this feature implemented? In fact, if you search the repository for NSScrollView, you are going to get quite a few hits in several files.

**Comment by cody271** [comment]
> Does this mean the MacOS backend already has a bit of this feature implemented?

Yes [`uiprivMkScrollView()`](https://github.com/libui-ng/libui-ng/blob/master/darwin/scrollview.m#L12) could be used here fairly easily on Darwin.

**Comment by petabyt** [comment]
Any idea on what the API for this should be?
```
uiScroll *uiNewScroll();
uiScrollSetChild(uiScroll *v, uiControl *child);
```
Or `uiScrollBox`? `uiScrollView`?

**Comment by petabyt** [comment]
win32 scrollable container. It's all messed up but it's there.
https://www.youtube.com/watch?v=Db-29q9Jq9c

Looks like a scrollable container on windows will have to have a subcontainer? Windows just gives you a scrollbar and makes you run everything manually. In this case the child of the `uiScroll` is being scrolled. This might not work for other widgets.

**Comment by swissChili** [comment]
Bump on this. If anybody is familiar with the OS X and windows ports and knows how to do this, this is a much needed feature. I wrote the old GTK PR but am not familiar with the other platforms.

**Comment by petabyt** [comment]
I made some half-baked Windows implementation here I think: https://github.com/petabyt/libui-cross/blob/master/extras/scroll.cpp
Very glitchy but it does work as you'd expect.

**Comment by petabyt** [comment]
> I made some half-baked Windows implementation here I think: https://github.com/petabyt/libui-cross/blob/master/extras/scroll.cpp Very glitchy but it does work as you'd expect.

If anybody can improve upon this, then we would pretty easily have `uiScroll` working on all platforms.

Which by the way, do we all agree on this API?
```
_UI_EXTERN uiScroll *uiNewScroll();
_UI_EXTERN void uiScrollSetChild(uiScroll *scroll, uiControl *ctl);
```
I chose `uiScroll` over `uiScrollView` to be consistent with `uiTab`.

**Comment by swissChili** [comment]
I tried to implement this on Mac, but ran into a lot of problems with actually getting the child to scroll (I have no experience with AppKit). It was not quite as easy as just using `uiprivMKScrollView`.

I can update my GTK fork to use this API, but I think somebody with more AppKit knowledge will need to do the Mac port.

**Comment by petabyt** [comment]
I think for Cocoa you just have to set the content size and it should work as is?
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/NSScrollViewGuide/Articles/Introduction.html#//apple_ref/doc/uid/TP40003460-SW1

I don't think it's like Windows where you have to listen to the scroll bar/scroll wheel events and scroll the contents yourself.
I would give it a MacOS port a shot as well, but I don't own a Mac.

**Comment by swissChili** [comment]
My [branch](https://github.com/swissChili/libui/tree/scroll-view) has what I've tried so far. The result is the child is inside the scroll view as expected, but doesn't actually scroll, it just gets clipped off at the top. Again, I don't know AppKit at all, so not sure what's going on.

<img width="438" alt="Screenshot 2024-05-28 at 8 40 42 PM" src="https://github.com/libui-ng/libui-ng/assets/26192527/ebe3d943-a2a9-4eaa-8a0e-a8a50e0b9bf2">

**Comment by swissChili** [comment]
Never mind, fixed it! Scrolling works on mac. I'll clean up and make a PR tomorrow.
<img width="432" alt="Screenshot 2024-05-28 at 8 51 45 PM" src="https://github.com/libui-ng/libui-ng/assets/26192527/5815a6d2-25f7-4fc7-b5db-32fe3c2c9dd1">

**Comment by petabyt** [comment]
Nice! I'll go ahead and clean up my Windows code and see if I can improve it.

**Comment by petabyt** [comment]
Draft PR: https://github.com/libui-ng/libui-ng/pull/280

**Comment by swissChili** [comment]
Unix and darwin: #281

**Comment by cblc** [comment]
Oh! That's great! I really needed this!!!
